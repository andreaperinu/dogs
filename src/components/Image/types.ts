import * as O from 'fp-ts/lib/Option'

export interface IImageProps {
	maybeImage: O.Option<string>
	fixedSize?: boolean
}
