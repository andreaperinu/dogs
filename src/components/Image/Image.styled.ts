import styled, { css } from 'styled-components';

export const ImageStyled = styled.img`
	display: block;
	margin: 3rem auto 1rem auto;
	width: 100%;
	max-width: 30rem;
`

export const FixedImageStyled = styled.div<{ imgSrc: string }>`
${({ imgSrc }) => css`
		display: block;
		margin: 3rem auto 1rem auto;
		width: 100%;
		max-width: 30rem;
		background: url(${imgSrc}) no-repeat center center; 
  	background-size: cover;
		padding-top: 56.25%
	`}
`
