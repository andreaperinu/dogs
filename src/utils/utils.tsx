import { sequenceT } from 'fp-ts/lib/Apply';
import * as A from 'fp-ts/lib/Array';
import { pipe } from 'fp-ts/lib/function';
import * as O from 'fp-ts/lib/Option';

import { ISelectItem } from '@/components/Select/types';
import { IBreed, IBreedResponseMessage } from '@/models/models';

export const noopComponent = () => <></>

export const maybeBreedToSelectItem = (maybeBreed: O.Option<IBreed>): O.Option<ISelectItem> => pipe(
	maybeBreed,
	O.map(({ breed }) => ({ id: breed, value: breed }))
)

export const maybeBreedsToSelectItems = (maybeBreeds: O.Option<IBreed[]>): O.Option<ISelectItem[]> => pipe(
	maybeBreeds,
	O.chain(O.fromPredicate(A.isNonEmpty)),
	O.map(A.map(({ breed }) => ({ id: breed, value: breed })))
)

export const maybeSubBreedToSelectItem = (subBreed: O.Option<string>): O.Option<ISelectItem> => pipe(
	subBreed,
	O.map(breed => ({ id: breed, value: breed })),
)

export const maybeSubBreedsToSelectItems = (breed: O.Option<IBreed>): O.Option<ISelectItem[]> => pipe(
	breed,
	O.map(({ subBreeds }) => subBreeds),
	O.map(A.map(value => ({ id: value, value: value })))
)

export const maybeBreedBySelectItem = (maybeBreeds: O.Option<IBreed[]>) => (
	maybeSelectItem: O.Option<ISelectItem>,
): O.Option<IBreed> => pipe(
	sequenceT(O.option)(maybeSelectItem, maybeBreeds),
	O.chain(([{ value }, breeds]) => A.findFirst(({ breed }: IBreed) => breed === value)(breeds))
)

export const maybeSubBreedBySelectItem = (
	maybeSelectItem: O.Option<ISelectItem>
): O.Option<string> => pipe(maybeSelectItem, O.map(({ value }) => value))

export const formatBreeds = (message: IBreedResponseMessage): IBreed[] => pipe(
	Object.keys(message),
	A.map(key => ({ breed: key, subBreeds: message[key] }))
)

export const filterBreedsBySubBreeds = (maybeBreeds: O.Option<IBreed[]>): O.Option<IBreed[]> => {

	const validSubBreedFilter = ({ subBreeds }: IBreed): boolean => pipe(subBreeds, A.isNonEmpty)

	return pipe(
		maybeBreeds,
		O.map(A.filter(validSubBreedFilter)),
	)
}
