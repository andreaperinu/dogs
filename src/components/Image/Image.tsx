import * as B from 'fp-ts/lib/boolean';
import { pipe } from 'fp-ts/lib/function';
import * as O from 'fp-ts/lib/Option';
import { memo, useCallback } from 'react';

import { noopComponent } from '@/utils/utils';

import { FixedImageStyled, ImageStyled } from './Image.styled';
import { IImageProps } from './types';

export const BaseImage = ({ maybeImage, fixedSize = false }: IImageProps) => {

	const renderImage = useCallback((src: string) => pipe(
		fixedSize,
		B.fold(
			() => <ImageStyled src={src} />,
			() => <FixedImageStyled imgSrc={src} />
		),
	), [fixedSize])

	return pipe(maybeImage, O.map(renderImage), O.getOrElse(noopComponent))
}

export const Image = memo(BaseImage)
