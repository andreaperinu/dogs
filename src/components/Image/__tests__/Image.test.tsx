import * as O from 'fp-ts/lib/Option';

import { setup } from '@/tests/utils';

import { Image as component } from '../Image';

const props: React.ComponentProps<typeof component> = {
	maybeImage: O.some('url')
}

describe('<Image />', () => {
	it('should match snapshot', () => {
		const wrapper = setup({ component, props, shouldMount: true })
		expect(wrapper).toMatchSnapshot()
	})

	it('should match snapshot (fixed)', () => {
		const wrapper = setup({ component, props: { ...props, fixedSize: true }, shouldMount: true })
		expect(wrapper).toMatchSnapshot()
	})
})
