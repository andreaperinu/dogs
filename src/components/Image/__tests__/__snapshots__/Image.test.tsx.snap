// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image /> should match snapshot (fixed) 1`] = `
<Memo(BaseImage)
  fixedSize={true}
  maybeImage={
    Object {
      "_tag": "Some",
      "value": "url",
    }
  }
>
  <styled.div
    imgSrc="url"
  >
    <div
      className="sc-gsTCUz jHqUdB"
    />
  </styled.div>
</Memo(BaseImage)>
`;

exports[`<Image /> should match snapshot 1`] = `
<Memo(BaseImage)
  maybeImage={
    Object {
      "_tag": "Some",
      "value": "url",
    }
  }
>
  <styled.img
    src="url"
  >
    <img
      className="sc-bdfBwQ gsGDQZ"
      src="url"
    />
  </styled.img>
</Memo(BaseImage)>
`;
